src/hera/workflows/_context.py

```diff
@@ -76,27 +76,29 @@ def add_sub_node(self, node: Union[SubNodeMixin, TTemplate]) -> None:
         if not pieces:
             return
 
-        try:
-            # here, we are trying to add a node to the last piece of context in the hopes that it is a subbable
-            pieces[-1]._add_sub(node)
-        except InvalidType:
-            # if the above fails, it means the user invoked a decorated function e.g. `@script`. Hence,
-            # the object needs to be added as a template to the piece of context at [-1]. This will be the case for
-            # DAGs and Steps
-            pieces[-1]._add_sub(node.template)  # type: ignore
-
-        # when the above does not raise an exception, it means the user invoked a decorated function e.g. `@script`
+        # for particular types, the user invoked a decorated function e.g. `@script`
         # inside a proper context. Here, we add the object to the overall workflow context, directly as a template,
         # in case it is not found (based on the name). This helps users save on the number of templates that are
         # added when using an object that is a `Script`
         if hasattr(node, "template") and node.template is not None and not isinstance(node.template, str):
             found = False
             for t in pieces[0].templates:  # type: ignore
                 if t.name == node.template.name:
+                    if t != node.template:
+                        raise TemplateNameConflict(f"Found multiple templates with the same name: {t.name}")
                     found = True
                     break
             if not found:
                 pieces[0]._add_sub(node.template)
 
+        try:
+            # here, we are trying to add a node to the last piece of context in the hopes that it is a subbable
+            pieces[-1]._add_sub(node)
+        except InvalidType:
+            # if the above fails, it means the user invoked a decorated function e.g. `@script`. Hence,
+            # the object needs to be added as a template to the piece of context at [-1]. This will be the case for
+            # DAGs and Steps
+            pieces[-1]._add_sub(node.template)  # type: ignore
```

Comment:
The old comment here didn't really make sense to me, and checking out the code it turns out it's never hit in our tests. cc @flaviuvadan I think the backend logic changed such that we don't need the try/except (i.e. callers of `_add_sub_node` always pass in a valid node type as it's called in `SubNodeMixin`, and I don't see a way to pass an invalid type). Tried removing it and tests all pass.

```suggestion
        pieces[-1]._add_sub(node)
```